### ISSUE TITLE
request: tqdm on kerchunk.hdf.SingleHdf5ToZarr.translate

### ISSUE THREAD

**Issue by bendichter** [fsspec/kerchunk#268]
running

```python
import kerchunk.hdf
import fsspec

s3_url =  "s3://dandiarchive/blobs/d6d/882/d6d88284-983a-42a7-91e8-48e9f59c2881"

with fsspec.open(
    s3_url,
    anon=True,
    default_fill_cache=False,
    default_cache_type='first',
) as inf:
    h5chunks = kerchunk.hdf.SingleHdf5ToZarr(inf, s3_url, inline_threshold=100)
    out = h5chunks.translate()
```

using the latest version of kerchunk, `kerchunk.hdf.SingleHdf5ToZarr.translate` is taking a long time for me (over 5 minutes so far) and has not yet completed. I may be using it incorrectly but if this is the intended usage it would be great to have the option of turning on a progress bar, e.g. tqdm, to track progress.

**Comment by martindurant** [comment]
You can at least turn on some logging to get feedback, e.g., 
```
fsspec.utils.setup_logging(logger_name="h5-to-zarr")
```
I don't know how much information you can get before starting about the total number of nodes to process - you would need that to make a progress bar worthwhile.

**Comment by bendichter** [comment]
@martindurant thanks this is helpful. Using the logger I can confirm that kerchunk is able to open the file and visit items, but is taking a long time to process the largest dataset. What about a per-dataset progress bar?

**Comment by martindurant** [comment]
You mean per HDF component? No, that sounds tricky.

Probably you want to play with the cache settings o see what works well. You can also get logging from the storage backend, "s3fs" in this case, which will tell you what calls are being made. Of course, the biggest improvement you can make is to run your kerchunking in AWS of the same region as the data.

**Comment by bendichter** [comment]
Thanks for the tips.

> Of course, the biggest improvement you can make is to run your kerchunking in AWS of the same region as the data.

I already am. I'm running on hub.dandiarchive.org accessing data on the DANDI Archive, which are in the same region.

How do I set up the logger on s3fs? I don't see anything about a logger in [the docs](https://s3fs.readthedocs.io/en/latest/api.html)

**Comment by martindurant** [comment]
```
fsspec.utils.setup_logging(logger_name="s3fs")
```

**Comment by bendichter** [comment]
oh, I see, thank you. I see now lots of calls to different bytes ranges so I suppose it is working as expected? Any tips of what to adjust with chunk settings to speed this up would be appreciated.

**Comment by martindurant** [comment]
I think the neuro HDF files have a much deeper nested tree structure than the typical netCDF4-like datasets which have just the one level and big arrays - so this is a bit of a test case. Fortunately, you only scan once to give access many times.

**Comment by bendichter** [comment]
That is true for some but not for this one- the vast majority of the data is in the dataset `/acquisition/ElectricalSeries/data`, which is about 70 GB. There is some other nested structure but the code is getting slowed down here.

**Comment by martindurant** [comment]
> Any tips of what to adjust with chunk settings to speed this up would be appreciated.

One useless tip: when creating the file, HDF can put the arrays in various orders. You will generally do best if the traversal of the tree (which is depth-first) matches the order in the file and the cache type is "readahead" with a decent block size. Likely, the traversal is actually random access, in which case readahead only adds overhead... The s3fs logging includes the fetched ranges, so that you can tell. Likely they are tiny reads in no particular order in the target file.

Unfortunately, we don't have a way to fetch the metadata pieces concurrently, which would be the very best thing to do here (lots of small reads -> bottleneck on latency). I don't think that h5py, the python driver library, can possibly do this for us. It would take someone re-engineering the HDF reader to know prospectively where all the metadata ranges are, which is surely a very tough task.

**Comment by martindurant** [comment]
Oh, and *sometimes* it's worth just downloading the file and scanning it locally. The hdf scanner allows you to specify where the file to read is and *also* where the eventual target will be.

**Comment by bendichter** [comment]
> One useless tip: when creating the file, HDF can put the arrays in various orders. You will generally do best if the traversal of the tree (which is depth-first) matches the order in the file and the cache type is "readahead" with a decent block size. Likely, the traversal is actually random access, in which case readahead only adds overhead... The s3fs logging includes the fetched ranges, so that you can tell. Likely they are tiny reads in no particular order in the target file.

That's interesting. Do you know how I might set up HDF5 to write this way in h5py?

> Unfortunately, we don't have a way to fetch the metadata pieces concurrently, which would be the very best thing to do here (lots of small reads -> bottleneck on latency). I don't think that h5py, the python driver library, can possibly do this for us. It would take someone re-engineering the HDF reader to know prospectively where all the metadata ranges are, which is surely a very tough task.

If this is much faster when the file is local this step may not be a bottleneck in the long run, if we make this part of the data ingestion process and perform the indexing either on the client side before upload or on the server side just after upload.

**Comment by ajelenak** [comment]
@bendichter If you have some time for testing, have access to `h5stat` and `h5repack` HDF5 CLI tools, and can put the new version of your HDF5 file back into S3, I can give you some tips that will probably significantly improve the performance. The issue here is that the internal file metadata are not cloud optimized.

**Comment by ajelenak** [comment]
> Do you know how I might set up HDF5 to write this way in h5py? 

Yes, you can. I added support for the file creation settings needed to produce cloud-optimized HDF5 files. Best is to use h5py from its master branch. `h5py.File(..., fs_strategy='page', fs_page_size=SIZE_IN_BYTES, ...)`. I don't see the h5py documentation with the complete info in either its latest or stable versions. So ping me if you are interested in this.

> If this is much faster when the file is local this step may not be a bottleneck in the long run, if we make this part of the data ingestion process and perform the indexing either on the client side before upload or on the server side just after upload.

Yes, that solves this particular slow down. But, your file would still come with chunks  small for cloud-type access. So it helps if the chunks are somewhere in the 4-16 MiB range. And this will help with reducing the total number of chunks so the kerchunk process is faster.

**Comment by bendichter** [comment]
@ajelenak very cool! So this feature hasn't been released on pypi/conda yet?

**Comment by ajelenak** [comment]
My bad... I forget when I add things in h5py... 😄 Looks like it's available since [v3.5](https://docs.h5py.org/en/latest/whatsnew/3.5.html). I also forgot to mention, if you do end up generating large chunks, you will need to increase dataset chunk cache when using such files. Those settings are available in h5py.
